pkg/wireguard/operator/operator.go

```diff
@@ -0,0 +1,369 @@
+// Copyright 2021 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// This package contains the WireGuard operator code, which is responsible to
+// allocate and set the WireGuard tunnel device IP address in the CiliumNode
+// object.
+package operator
+
+import (
+	"errors"
+	"fmt"
+	"net"
+
+	v2 "github.com/cilium/cilium/pkg/k8s/apis/cilium.io/v2"
+	"github.com/cilium/cilium/pkg/lock"
+	"github.com/cilium/cilium/pkg/logging"
+	"github.com/cilium/cilium/pkg/logging/logfields"
+	"github.com/cilium/cilium/pkg/node/addressing"
+	"github.com/cilium/cilium/pkg/option"
+	"github.com/cilium/cilium/pkg/wireguard/types"
+
+	"github.com/cilium/ipam/service/ipallocator"
+	"github.com/sirupsen/logrus"
+	"k8s.io/client-go/util/retry"
+)
+
+var log = logging.DefaultLogger.WithField(logfields.LogSubsys, "wireguard")
+
+type CiliumNodeUpdater interface {
+	Update(origNode, node *v2.CiliumNode) (*v2.CiliumNode, error)
+	Get(node string) (*v2.CiliumNode, error)
+}
+
+type Operator struct {
```

Comment:
Thanks for the well-documented code! I enjoyed reading it!

How about adding something like the following to provide a basic understanding of restoring?
```
// Operator is initialized with NewOperator()
//
// The operator starts in restoring mode, when it will not attempt to allocate
// any IPs and, instead, place nodes in ipv{4,6}AllocAfterRestore so that their
// IPs can be allocated on RestoreFinished()
```